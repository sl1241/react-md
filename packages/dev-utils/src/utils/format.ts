import log from "loglevel"
import type { BuiltInParserName } from "prettier"
import prettier from "prettier"

function getParser(
    code: string,
    parser: BuiltInParserName | undefined,
    filePath: string | undefined
): BuiltInParserName | undefined {
    if (typeof parser === "string") {
        return parser
    }

    if (typeof filePath === "string") {
        return undefined
    }

    if (code.startsWith("{") || code.startsWith("[")) {
        return "json"
    }

    if (/^(@import|\.[a-z])/m.test(code) || /^\s+@(mixin|include)/.test(code)) {
        return "scss"
    }

    return "typescript"
}

export function format(
    code: string,
    parser?: BuiltInParserName,
    filepath?: FunctionMaybe<Nullable<string>>
): string {
    const resolvedParser = getParser(code, parser, filepath)
    try {
        return prettier.format(code, {
            filepath,
            parser: resolvedParser,
            proseWrap: resolvedParser === "markdown" ? "always" : undefined,
            singleQuote: resolvedParser === "scss" || resolvedParser === "babel",
        })
    } catch (e) {
        log.error(e)
        process.exit(1)
    }
}
