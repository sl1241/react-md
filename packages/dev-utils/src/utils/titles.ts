import { upperFirst } from "lodash"

/**
 * A simple utility that converts some string into a page/section title. This
 * really just capitalizes each word in the string after it has been split by
 * hyphens or capital letters.
 *
 * @param s - The string to convert to a title.
 * @param joinWith - A string that should be used to join all the parts together
 * to create a title.
 * @param capitals - Boolean if the string should be split on capital letters
 * instead of hyphens.
 * @returns A title-ized string.
 */
export function toTitle(s: string, joinWith = " ", capitals = false): string {
    if (/autocomplete/i.test(s)) {
        return "AutoComplete"
    }

    if (/^api$/i.test(s)) {
        return "API"
    }

    if (/sassdoc/i.test(s)) {
        return "SassDoc"
    }

    if (s === "[id]") {
        // TODO: Fix to use query params instead
        return "Layout"
    }

    return s
        .split(capitals ? /(?=[A-Z])/ : "-")
        .map((part) => {
            if (/^((v\d+)|(to))$/.test(part)) {
                return part
            }

            if (/^api$/i.test(part)) {
                return "API"
            }

            if (/^cdn$/i.test(part)) {
                return "CDN"
            }

            return upperFirst(part)
        })
        .join(joinWith)
}

/**
 * Converts a pathname into a breadcrumb type of title to display in the
 * documentation site.
 *
 * @param pathname - The current pathname to create a title for
 * @param statusCode - An optional status code to check against for custom error
 * cases.
 * @returns a string to display in the documentation site's breadcrumbs.
 */
export function toBreadcrumbPageTitle(
    pathname: string,
    statusCode?: FunctionMaybe<Nullable<number>>
): string {
    let title = ""
    if (statusCode) {
        switch (statusCode) {
            case 404:
                title = "Not Found"
                break
            default:
                title = "Server error"
        }
    } else if (/v\d+-to-v\d+$/.test(pathname)) {
        const [migration] = pathname.split("/").reverse()
        title = `Migration Guides - ${migration.replace(/-/g, " ")}`
    } else {
        const parts = pathname.split("/").filter((p) => !!p && !/packages/.test(p))
        title = parts.map((p) => toTitle(p)).join(" - ")
    }

    return `react-md${title ? ` - ${title}` : ""}`
}

/**
 * This is a utility function that will create a unique id for a title string.
 * The title string should normally be somewhere betwen 5-20 characters.
 *
 * This is generally used to create ids for links within the documentation page
 * for headers to generate quick links and table of contents.
 *
 * @param title - The title to create an id for.
 * @returns an id based on the title
 */
export function toId(title: string): string {
    return title
        .replace(/\/|\\|\[|]|-/g, "")
        .replace(/SCSS/g, "scss ")
        .replace(/CSS/g, "css ")
        .replace(/API/g, "api")
        .replace(/\s+/g, " ")
        .trim()
        .split(/\s|(?=[A-Z])/)
        .join("-")
        .toLowerCase()
}
