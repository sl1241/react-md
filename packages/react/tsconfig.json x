{
  "extends": "../../tsconfig.base.json",
  "compilerOptions": {
    "outDir": "./dist",
    "rootDir": "./src",
    "declarationDir": "./dist/types",
    "composite": true,
    "skipLibCheck": true,
    "skipDefaultLibCheck": true,
    "jsx": "preserve",
    "jsxImportSource": "voby",
    "types": [
      "voby/dist/jsx/jsx"
    ],
  },
  "exclude": [
    "vite.config.ts",
"vite.config.umd.ts",
    "types",
  ],
}
