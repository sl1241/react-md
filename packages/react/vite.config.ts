import { defineConfig } from 'vite'

const config = defineConfig({
    build: {
        minify: false,
        lib: {
            entry: ["./src/index.ts"],
            name: "react",
            formats: ['cjs', 'es', 'umd'],
            fileName: (format: string, entryName: string) => `${entryName}.${format}.js`
        },
        // rollupOptions: {
        //   output: {
        //     preserveModules: true,
        //   }
        // },
    },


    esbuild: {
        jsx: 'automatic',
    },

    css: {
        modules: {
            scopeBehaviour: 'global',
            localsConvention: 'camelCase',
        }
    },
})



export default config
