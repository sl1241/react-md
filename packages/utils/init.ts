import '@testing-library/jest-dom'
import '@testing-library/jest-dom/extend-expect'

// import {
//   DEFAULT_DESKTOP_MIN_WIDTH,
//   DEFAULT_DESKTOP_LARGE_MIN_WIDTH,
// } from '../packages/utils/lib/sizing/constants'

// // window will be undefined for the one test I force to be run in node instead
// // of jsdom
// if (typeof window !== 'undefined' && typeof window.matchMedia !== 'function') {
//   window.matchMedia = (query) => ({
//     media: query,
//     matches:
//       query.includes(`${DEFAULT_DESKTOP_MIN_WIDTH}`) ||
//       query.includes(`${DEFAULT_DESKTOP_LARGE_MIN_WIDTH}`),
//     onchange: () => {},
//     addListener: () => {},
//     removeListener: () => {},
//     addEventListener: () => {},
//     removeEventListener: () => {},
//     dispatchEvent: () => false,
//   });
// }
