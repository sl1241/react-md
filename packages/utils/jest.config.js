/** @type {import('ts-jest').JestConfigWithTsJest} */
// module.exports = {
//     preset: 'ts-jest',
//     testEnvironment: 'jsdom',

//     // globals: {
//     //     'ts-jest': {
//     //         compiler: "typescript"
//     //         // ts-jest configuration goes here
//     //     }
//     // },

//     transform: {
//         "^.+\\.tsx?$": "ts-jest",
//     },
//     testRegex: "(/__tests__/.*|(\\.|/)(test|spec))\\.(jsx?|tsx?)$",
//     moduleFileExtensions: ["ts", "tsx", "js", "jsx", "json", "node"],
// }

module.exports = {
  preset: 'ts-jest',
  testEnvironment: 'jsdom',

  setupFilesAfterEnv: ['<rootDir>/init.ts'],

  testMatch: ['**/?(*.)+(spec|test).ts?(x)', '**/__tests__/**/*.ts?(x)'],
  globals: {
    'ts-jest': {
      tsconfig: {
        jsx: 'react-jsx',
        jsxImportSource: 'voby',
        strict: true,
        moduleResolution: 'node',
        // preserveSymlinks: true,
        noImplicitAny: false,
        strictNullChecks: false,
        lib: ['dom', 'dom.iterable', 'esnext'],
        types: ['voby/dist/jsx/jsx', 'jest'],
      },
    },
  },
  // transform: {
  //     // '\\.svg$': '<rootDir>/testSetup/inlineSvgTransformer.js',
  //     // '\\.(jpg|jpeg|png|gif|eot|otf|webp|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$':
  //     //     '<rootDir>/testSetup/fileTransformer.js',
  //     '^.+\\.tsx?$': ['ts-jest', {
  //         // tsconfig: './tsconfig.test.json',
  //         // useESM: true,
  //         // // other ts-jest config options
  //         jsx: 'react-jsx',
  //         jsxImportSource: 'voby/testing',
  //         strict: true,
  //         moduleResolution: 'node',
  //         // preserveSymlinks: true,
  //         noImplicitAny: false,
  //         strictNullChecks: false,
  //         lib: [
  //             'dom',
  //             'dom.iterable',
  //             'esnext',
  //         ],
  //         types: [
  //             'voby/dist/jsx/jsx',
  //             'jest',
  //         ],
  //     }],
  // },

  moduleNameMapper: {
    '^@/(.*)$': '<rootDir>/src/$1',
  },
  moduleFileExtensions: ['js', 'json', 'jsx', 'node', 'ts', 'tsx'],
};
