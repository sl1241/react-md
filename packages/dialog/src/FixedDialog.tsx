import type { CSSProperties, ObservableMaybe } from 'voby'

import type { CSSTransitionClassNames } from "@react-md/transition"
import { useFixedPositioning } from "@react-md/transition"
import type {
  CalculateFixedPositionOptions,
  LabelRequiredForA11y,
} from "@react-md/utils"
import { TOP_INNER_RIGHT_ANCHOR } from "@react-md/utils"

import type { BaseDialogProps } from "./Dialog"
import { Dialog } from "./Dialog"

export interface BaseFixedDialogProps
  extends Omit<BaseDialogProps, "type">,
  Pick<CalculateFixedPositionOptions, "anchor"> {
  /**
   * The element the dialog should be fixed to. This can either be:
   * - a query selector string to get an element
   * - an HTMLElement (normally a ref.current)
   * - a function that returns an HTMLElement or null
   * - null
   */
  fixedTo: ObservableMaybe<HTMLElement>

  /**
   * Any additional options to apply to the fixed positioning logic. The
   * `transformOrigin` option will be enabled by default.
   */
  options?: FunctionMaybe<CalculateFixedPositionOptions>

  /**
   * An optional function to call to get the fixed positioning options.
   */
  getOptions?(): CalculateFixedPositionOptions
}

export type FixedDialogProps = LabelRequiredForA11y<BaseFixedDialogProps>

const DEFAULT_CLASSNAMES: CSSTransitionClassNames = {
  appear: "rmd-dialog--fixed-enter",
  appearActive: "rmd-dialog--fixed-enter-active",
  enter: "rmd-dialog--fixed-enter",
  enterActive: "rmd-dialog--fixed-enter-active",
  exit: "rmd-dialog--fixed-exit",
  exitActive: "rmd-dialog--fixed-exit-active",
}

/**
 * The `FixedDialog` is a wrapper for the `Dialog` component that can be used to
 * be fix itself to another element. Another term for this component might be a
 * "Pop out Dialog".
 */
export const FixedDialog = (
  {
    fixedTo,
    style: propStyle,
    anchor = TOP_INNER_RIGHT_ANCHOR,
    options,
    getOptions,
    children,
    className,
    classNames = DEFAULT_CLASSNAMES,
    overlayHidden = true,
    disableScrollLock = true,
    onEnter,
    onEntering,
    onEntered,
    onExited,
    ref: nodeRef,
    ...props
  }: FixedDialogProps //& { style: FunctionMaybe<Nullable<string | StyleProperties>>, children?: Children, className?: Class, ref: ObservableMaybe<HTMLDivElement> },
) => {
  const { onRequestClose } = props

  const { ref, style, callbacks } = useFixedPositioning({
    //@ts-ignore
    nodeRef,
    style: propStyle,
    transformOrigin: true,
    onEnter,
    onEntering,
    onEntered,
    onExited,
    anchor,
    fixedTo,
    onScroll: /* istanbul ignore next */ (_event, { visible }) => {
      if (!visible) {
        onRequestClose()
      }
    },
    ...options,
    getFixedPositionOptions: getOptions,
  })

  //@ts-ignore
  return <Dialog
    {...props}
    {...callbacks}
    //@ts-ignore
    ref={ref}
    type="custom"
    //@ts-ignore
    style={style}
    className={["rmd-dialog--fixed", className]}
    classNames={classNames}
    overlayHidden={overlayHidden}
    disableScrollLock={disableScrollLock}
  >
    {children}
  </Dialog>
}
