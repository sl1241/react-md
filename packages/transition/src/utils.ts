import { getViewportSize } from "@react-md/utils"
import type {
    CSSTransitionClassNames,
    CSSTransitionClassNamesObject,
    TransitionActions,
    TransitionTimeout,
    TransitionTimeoutObject,
} from "./types"
import { CSSProperties, $$ } from "voby"

/**
 * @remarks \@since 4.0.0
 * @internal
 */
export interface TransitionTimeoutOptions extends Required<TransitionActions> {
    timeout: FunctionMaybe<TransitionTimeout>
}

/**
 * @remarks \@since 4.0.0
 * @internal
 */
export function getTransitionTimeout({ timeout, appear, enter, exit, }: Readonly<TransitionTimeoutOptions>): Readonly<Required<TransitionTimeoutObject>> {
    const to = $$(timeout)

    if (typeof to === "number") {
        return {
            appear: appear ? to : 0,
            enter: enter ? to : 0,
            exit: exit ? to : 0,
        }
    }

    return {
        appear: (appear && (to.appear ?? to.enter)) || 0,
        enter: (enter && to.enter) || 0,
        exit: (exit && to.exit) || 0,
    }
}

/**
 * @remarks \@since 4.0.0
 * @internal
 */
export interface CollapseSizing {
    maxHeight?: CSSProperties['maxHeight']
    paddingTop?: CSSProperties['paddingTop']
    paddingBottom?: CSSProperties['paddingBottom']
}

/**
 * A small util that will find the max-height, padding-top, and padding-bottom
 * for the provided element. This is really used to be able to transition the
 * max-height value since `max-height: auto` does not transition. The only way
 * to get transition is to change max-height values manually.
 *
 * @remarks \@since 4.0.0
 * @internal
 */
export function getElementSizing(element: HTMLElement | null): CollapseSizing {
    let maxHeight
    let paddingTop
    let paddingBottom
    if (element) {
        maxHeight = element.scrollHeight

        // clone the element (not deep) just to figure out it's padding without the
        // inline styles applied
        const cloned = element.cloneNode(false) as HTMLElement
        cloned.style.maxHeight = ""
        cloned.style.padding = ""
        cloned.style.paddingLeft = element.style.paddingLeft
        cloned.style.paddingRight = element.style.paddingRight
        cloned.style.visibility = "hidden"

        const container = element.parentElement || document.body
        container.appendChild(cloned)
        const style = window.getComputedStyle(cloned)
        if (style.paddingTop) {
            paddingTop = parseFloat(style.paddingTop)
        }

        if (style.paddingBottom) {
            paddingBottom = parseFloat(style.paddingBottom)
        }
        container.removeChild(cloned)
    }

    return { maxHeight, paddingTop, paddingBottom }
}

/**
 * @remarks \@since 4.0.0
 * @internal
 */
export interface TransitionClassNamesOptions extends TransitionTimeoutOptions {
    classNames: FunctionMaybe<CSSTransitionClassNames>
}

/**
 * @remarks \@since 4.0.0
 * @internal
 */
export function getTransitionClassNames({ classNames: cs, ...timeoutOptions }: TransitionClassNamesOptions): Readonly<Required<CSSTransitionClassNamesObject>> {
    const classNames = $$(cs)
    const timeout = getTransitionTimeout(timeoutOptions)

    if (typeof classNames === "string") {
        const { appear, enter, exit } = timeout
        return {
            appear: appear ? `${classNames}--appear` : "",
            appearActive: appear ? `${classNames}--appear-active` : "",
            appearDone: "",
            enter: enter ? `${classNames}--enter` : "",
            enterActive: enter ? `${classNames}--enter-active` : "",
            enterDone: "",
            exit: exit ? `${classNames}--exit` : "",
            exitActive: exit ? `${classNames}--exit-active` : "",
            exitDone: "",
        }
    }

    const {
        enter = "",
        enterActive = "",
        enterDone = "",
        exit = "",
        exitActive = "",
        exitDone = "",
        appear = (timeout.appear && enter) || "",
        appearActive = (timeout.appear && enterActive) || "",
        appearDone = (timeout.appear && enterDone) || "",
    } = classNames

    return {
        appear,
        appearActive,
        appearDone,
        enter,
        enterActive,
        enterDone,
        exit,
        exitActive,
        exitDone,
    }
}

/**
 * @remarks \@since 4.0.0
 * @internal
 */
interface IsWithinViewportOptions {
    fixedElement: HTMLElement
    fixedToElement: HTMLElement
}

/**
 * @remarks \@since 4.0.0
 * @internal
 */
export function isWithinViewport({
    fixedElement,
    fixedToElement,
}: IsWithinViewportOptions): boolean {
    const fixedElementRect = fixedElement.getBoundingClientRect()
    const fixedToElementRect = fixedToElement.getBoundingClientRect()
    const vh = getViewportSize("height")
    const vw = getViewportSize("width")
    const top = Math.min(fixedElementRect.top, fixedToElementRect.top)
    const right = Math.max(fixedElementRect.right, fixedToElementRect.right)
    const bottom = Math.max(fixedElementRect.bottom, fixedToElementRect.bottom)
    const left = Math.min(fixedElementRect.left, fixedToElementRect.left)

    return bottom >= 0 && top <= vh && right >= 0 && left <= vw
}
