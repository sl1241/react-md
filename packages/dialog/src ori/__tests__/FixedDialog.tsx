import type { ReactElement } from "react";
import { useRef } from "react";
import { render } from "@testing-library/react";
import { DATA_RMD_NOSCROLL } from "@react-md/utils";

import type { FixedDialogProps } from "../FixedDialog";
import { FixedDialog } from "../FixedDialog";

const noop = (): void => {};

function Test({
  id = "dialog",
  "aria-label": ariaLabel = "Label",
  onRequestClose = noop,
  visible = false,
  children,
  ...props
}: Partial<FixedDialogProps>): ReactElement {
  const fixedTo = useRef<HTMLButtonElement>(null);
  return (
    <>
      <button type="button" ref={fixedTo}>
        Fixed To
      </button>
      <FixedDialog
        {...props}
        id={id}
        aria-label={ariaLabel}
        visible={visible}
        onRequestClose={onRequestClose}
        fixedTo={fixedTo}
      >
        {children}
      </FixedDialog>
    </>
  );
}

describe("FixedDialog", () => {
  it("should disable the overlay and scroll lock behavior by default", () => {
    const onRequestClose = jest.fn();
    const { baseElement } = render(
      <Test visible onRequestClose={onRequestClose} />
    );

    expect(baseElement).toMatchSnapshot();
    expect(document.body).not.toHaveAttribute(DATA_RMD_NOSCROLL);
    expect(onRequestClose).not.toBeCalled();
  });
});
