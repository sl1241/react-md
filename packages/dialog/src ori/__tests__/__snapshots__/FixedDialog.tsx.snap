// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FixedDialog should disable the overlay and scroll lock behavior by default 1`] = `
<body>
  <div>
    <button
      type="button"
    >
      Fixed To
    </button>
  </div>
  <span
    class="rmd-overlay rmd-overlay--visible rmd-overlay--clickable rmd-dialog-overlay"
    id="dialog-overlay"
    tabindex="-1"
  />
  <div
    aria-label="Label"
    aria-modal="true"
    class="rmd-dialog rmd-dialog--fixed"
    id="dialog"
    role="dialog"
    tabindex="-1"
  />
</body>
`;
